### PR TITLE
[Messenger] Envelope as first class citizen in middleware too

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | 4.2 
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | yes     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets |  N/A  
| License       | MIT
| Doc PR        | N/A

Nothing really new here, as I already suggested doing so in the original PR introducing the `Envelope` object, but we wanted first to give a try about keeping the ability to just get the message in middleware.

Now, I'm still convinced making the Envelope a first-class citizen for everything inside the bus would homogenize and simplify things for everyone and is a better design than the `EnvelopeAwareInterface`, while keeping the message at both ends of the bus (i.e: `MessageBusInterface::dispatch` and handlers).

Cons:

- Not what command buses' users are already used to
- Requires an extra call to `$envelope->getMessage()` in a middleware not using the `Envelope`
- `Envelope` seems not really relevant in some parts of a CQRS / layered architecture app (...or is it?)

Pros:

- Homogeneous use of the `Envelope` anywhere inside the bus
- Strict & faithful `MiddlewareInterface::handle(Envelope $envelope)` typehint.
  Not cheating with the non-restrictive `object` typehint which was never intended for this feature.
- No more `EnvelopeAwareInterface` as there is no more core usage
- No more manipulation regarding the next message or envelope consumer:
  e.g:
    - base message bus implementation is simplified
    - middleware decorators won't have to care about the inner middleware type with the `Envelope::getMessageFor($target)` util ([`ActivationMiddlewareDecorator`](https://github.com/symfony/symfony/pull/27320), [`TraceableMiddleware`](https://github.com/symfony/symfony/pull/27321), ...). Currently it means such decorators must absolutely implement `EnvelopeAwareInterface` and care about this. Also see https://github.com/symfony/symfony-docs/pull/9757#discussion_r189838501

My opinion is the naming and the concepts introduced by the `Envelope` VO are small and abstract (envelope, message, items) enough to even fit in the CQRS world or in a middleware in my `Application` namespace (when such a middleware is relevant).

---

img_0835
(_cheers from Spain_)

### CODE REVIEW

**Review by yceruto — COMMENTED** [review]
This looks better to me, thank you!

File: src/Symfony/Component/Messenger/Tests/Asynchronous/Middleware/SendMessageMiddlewareTest.php [hunk 1]
```diff
@@ -53,78 +39,78 @@ public function testItSendsTheMessageToAssignedSenderWithPreWrappedMessage()
 
     public function testItAlsoCallsTheNextMiddlewareBasedOnTheMessageClass()
     {
-        $message = new DummyMessage('Hey');
+        $envelope = Envelope::wrap(new DummyMessage('Hey'));
```
[the remark this answers is not in the document]
Comment (ogizanagi) [reply src/Symfony/Component/Messenger/Tests/Asynchronous/Middleware/SendMessageMiddlewareTest.php]: You're right. I think I'd remove this named construct as well as it doesn't really serve a purpose anymore, appart the fluent api without extra `(` `)`

**Comment by sroze** [comment]
Let's move this to 4.2; time for 4.1-beta2 :)

**Comment by hhamon** [comment]
Postponing to 4.2 means having a BC break.

**Comment by jvasseur** [comment]
@hhamon BC breaks are OK since the component is experimental

**Comment by ogizanagi** [comment]
Right, I’ve added the BC break label already. I should have submitted this sooner, my bad. The component is experimental so we could just document the upgrade path in 4.2 but I agree this is far from being ideal to require such a change for end-users and lib maintainers.
I already agreed with @sroze to move it to 4.2. There is only a few time before 4.1 release and if this is not part of beta2 we can’t make it. However if we don’t want this now, will it ever happen anyway?

**Comment by ogizanagi** [comment]
If that's the only fear you have about this, I'd suggest to add an abstract `Middleware` class which would just ask users to implement a `handleMessage(object $message)` instead (in the same spirit of the abstract `Voter` class hiding some complexity from end-users).
To me, comments & facts like in https://github.com/symfony/symfony-docs/pull/9757#discussion_r187367461 hint this design is a bit flawed this way.

**Comment by sroze** [comment]
Following #27841 being merged, I'm closing this one :)

**Comment by ogizanagi** [comment]
#27841 being merged does not change anything regarding this PR actually. 
But we've already expressed each other our arguments multiple times. I would have expected to get more insights from the community after using the component and this specific feature (still too soon?), this PR allowing discussions. So, your call. 🤷‍♂️

**Pull request closed without merging**